google-cloud-clients/google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java

```diff
@@ -889,95 +893,184 @@ public void onEvent(
-                            assertEquals(Type.ADDED, value.getDocumentChanges().get(0).getType());
-                            ref1.set(map("foo", "bar"));
-                            break;
-                          case 2:
-                            actualEvents.add("event 2");
-                            assertEquals(2, value.size());
-                            assertEquals(1, value.getDocumentChanges().size());
-                            assertEquals(Type.ADDED, value.getDocumentChanges().get(0).getType());
-                            ref1.set(map("foo", "bar", "bar", " foo"));
-                            break;
-                          case 3:
-                            actualEvents.add("event 3");
-                            assertEquals(2, value.size());
-                            assertEquals(1, value.getDocumentChanges().size());
-                            assertEquals(
-                                Type.MODIFIED, value.getDocumentChanges().get(0).getType());
-                            ref2.set(map("foo", "foo"));
-                            break;
-                          case 4:
-                            actualEvents.add("event 4");
-                            assertEquals(1, value.size());
-                            assertEquals(1, value.getDocumentChanges().size());
-                            assertEquals(Type.REMOVED, value.getDocumentChanges().get(0).getType());
-                            ref1.delete();
-                            break;
-                          case 5:
-                            actualEvents.add("event 5");
-                            assertTrue(value.isEmpty());
-                            assertEquals(1, value.getDocumentChanges().size());
-                            assertEquals(Type.REMOVED, value.getDocumentChanges().get(0).getType());
-                            break;
-                        }
-                      } catch (Exception e) {
-                        fail(e.getMessage());
-                      }
-                      semaphore.release();
-                    }
-                  });
-
-      final boolean tryAcquire = semaphore.tryAcquire(6, 60, TimeUnit.SECONDS);
-
-      final Joiner j = Joiner.on(", ");
-      final String expectedString = j.join(expectedEvents);
-      final String actualString = j.join(actualEvents);
-      assertTrue(
-          String.format(
-              "did not receive all expected events within the deadline.%n"
-                  + "expectedEvents = [%s]%n"
-                  + "  actualEvents = [%s]%n",
-              expectedString, actualString),
-          tryAcquire);
+          query.addSnapshotListener(
+              new EventListener<QuerySnapshot>() {
+                @Override
+                public void onEvent(
+                    @Nullable QuerySnapshot value, @Nullable FirestoreException error) {
+                  receivedEvents.add(new ListenerEvent(value, error));
```

Comment:
We should never see an `error` here. To simplify this test, I would add an assert here that the error is null (and maybe that `value` is not) and replace ListenerEvent with QuerySnapshot.

Reply:
Asserts inside the snapshotListener unfortunately do not automatically cascade outside the listener and as such can't fail the test, that I why I pulled it up to the top level and added the assertion after all the operations.

Reply:
What happens if we don't see the right number of snapshots (we get an error but we ignore it and don't add a receivedEvent)? My only reservation with this PR is that the generic nature of it adds a lot of code that makes the event flow somewhat hard to parse. I'm totally biased here, but the original test seems significantly easier to follow.

Reply:
When I started investigating why this tests would fail ~65% of the time in CI I narrowed it down to the fact that the semaphore wasn't ever being satisfied. In #5746 I added a timeout as well as some diagnostic information to the test and found that the failure always occurred in the `case 0` brach and resulted in the other branches never being executed. To ensure that performing new document updates wouldn't ever block the processing of received I separated the update and the lister as presented in this PR. I have ran this new implementation continuously for 2,500 iterations and all 2,500 runs passed.

I agree that it is more code -- a fair amount of it would be reduced if we were able to use java 1.8 constructs (which will hopefully happen sooner rather than later) -- and not super straightforward from a cursory glance, but I don't think the original test was either. One advantage with this new implementation is that it's more explicit about what the interrelationships are between the listener and document updates. In addition, there are explicit timeouts defined for each type of operation so we don't ever hang the entire build, and having the assertions outside the listeners means if there is a failure it will actually fail the test.

If it isn't important for `doc1` and `doc2` to be updated independent from one another I can merge those workers. If we want to avoid the anonymous inner classes from `FluentIterable` I can rewrite that with for loops.